### DISCUSSION TITLE
Low PUT performance with enabled write cache

### DISCUSSION THREAD

**Discussion by diawayeu** [minio/minio#13903] [Q&A] [answered]
We tested a standalone node configuration and noticed that the PUT speed is slow. Adding an NVMe cache does not significantly affect the results (see the following test 1 vs tests 2 & 3).

**Linux:**
Fedora release 36 (Rawhide) kernel 5.16.0

**MinIO Version:**
Version: 2021-12-10T23:03:39Z

**Used drives:**
8 x HDD (SKU: WDC WUH722020ALE6L4)
4 x NVMe (SKU: WUS4BB096D7P3E4)

**File system:**
```
mkfs.xfs /dev/sd${disk}
mount /dev/sd${disk} /mnt/minio/disk${num}

mkfs.xfs /dev/nvme${disk}n1
mount /dev/nvme${disk}n1 /mnt/minio/cache${num}  
```
  
**System information:**
CPU: AMD EPYC 7313 16-Core Processor
RAM: 8 x 32GB

We used the s3-benchmark testing tool provided by Wasabi (https://github.com/wasabi-tech/s3-benchmark).

**Test 1: 8xHDD (without cache)**

Environment variables:
export MINIO_STORAGE_CLASS_STANDARD=EC:2

Command start minio server:
./minio server /mnt/minio/disk{1...8} --console-address ":9001"

```
# ./s3-benchmark -u http://XXX.XXX.XXX.XXX:9000/test -a minio -s minio123 -t 5 -z 1M -d 60
Wasabi benchmark program v2.0
Parameters: url=http://XXX.XXX.XXX.XXX:9000/test, bucket=wasabi-benchmark-bucket, region=us-east-1, duration=60, threads=5, loops=1, size=1M
Loop 1: PUT time 60.5 secs, objects = 525, speed = 8.7MB/sec, 8.7 operations/sec. Slowdowns = 0
Loop 1: GET time 60.0 secs, objects = 443565, speed = 7.2GB/sec, 7392.7 operations/sec. Slowdowns = 0
Loop 1: DELETE time 0.3 secs, 1757.5 deletes/sec. Slowdowns = 0
```

**Test 2: 8xHDD + Cache from 4xNVMe**

Environment variables:
export MINIO_STORAGE_CLASS_STANDARD=EC:2
export MINIO_CACHE="on"
export MINIO_CACHE_DRIVES="/mnt/minio/cache{1...4}"
export MINIO_CACHE_QUOTA=80
export MINIO_CACHE_AFTER=0
export MINIO_CACHE_WATERMARK_LOW=70
export MINIO_CACHE_WATERMARK_HIGH=90

Command start minio server:
./minio server /mnt/minio/disk{1...8} --console-address ":9001"

```
# ./s3-benchmark -u http://XXX.XXX.XXX.XXX:9000/test -a minio -s minio123 -t 5 -z 1M -d 60
Wasabi benchmark program v2.0
Parameters: url=http://XXX.XXX.XXX.XXX:9000/test, bucket=wasabi-benchmark-bucket, region=us-east-1, duration=60, threads=5, loops=1, size=1M
Loop 1: PUT time 61.0 secs, objects = 518, speed = 8.5MB/sec, 8.5 operations/sec. Slowdowns = 0
Loop 1: GET time 60.0 secs, objects = 396052, speed = 6.4GB/sec, 6600.8 operations/sec. Slowdowns = 0
Loop 1: DELETE time 0.5 secs, 1131.5 deletes/sec. Slowdowns = 0
```

**Test 3: 8xHDD + Cache from 4xNVMe (Set to 'writethrough' for upload caching)**

Environment variables:
export MINIO_STORAGE_CLASS_STANDARD=EC:2
export MINIO_CACHE="on"
export MINIO_CACHE_DRIVES="/mnt/minio/cache{1...4}"
export MINIO_CACHE_QUOTA=80
export MINIO_CACHE_AFTER=0
export MINIO_CACHE_WATERMARK_LOW=70
export MINIO_CACHE_WATERMARK_HIGH=90
export MINIO_CACHE_COMMIT="writethrough"

Command start minio server:
./minio server /mnt/minio/disk{1...8} --console-address ":9001"

```
# ./s3-benchmark -u http://XXX.XXX.XXX.XXX:9000/test -a minio -s minio123 -t 5 -z 1M -d 60
Wasabi benchmark program v2.0
Parameters: url=http://XXX.XXX.XXX.XXX:9000/test, bucket=wasabi-benchmark-bucket, region=us-east-1, duration=60, threads=5, loops=1, size=1M
Loop 1: PUT time 60.2 secs, objects = 587, speed = 9.8MB/sec, 9.8 operations/sec. Slowdowns = 0
Loop 1: GET time 60.0 secs, objects = 392459, speed = 6.4GB/sec, 6540.9 operations/sec. Slowdowns = 0
Loop 1: DELETE time 0.5 secs, 1116.3 deletes/sec. Slowdowns = 0
```

I tried optimizations as per the guide https://min.io/resources/docs/MinIO-Throughput-Benchmarks-on-HDD-24-Node.pdf

Are the received numbers in line with what one can expect from one node with NVMe cache or something fundamental is missed?

Please advise.

**Comment by klauspost** [comment, marked as answer]
Caching is not something we generally recommend. It doesn't work in distributed setups and `writethrough` forces writes to be persisted on the backend, so your cache does nothing for PUT. `writeback` is only recently supported and generally not recommended anyway. But you can of course try.

It seems like GET requests are just reading from system cache, so it doesn't really provide much information.

You can look into tiering: https://docs.min.io/minio/baremetal/lifecycle-management/transition-objects-to-s3.html#

This will keep your NVME as "front line", storing objects only there and tiering the data to slower storage after a while. Metadata remains on NVME.

But even tiering requires your "slow" disks to be able to keep up with ingress, to avoid filling NVME, so several factors are needed to evaluate.

**Comment by diawayeu** [comment]
We looked at the transitional objects.
It can be seen from the documentation that the transition from NVMe storage to HDD storage can be specified at least one calendar day.

```
mc ilm add ALIAS/BUCKET \
--storage-class TIERNAME \
--transition-days DAYS \
--noncurrentversion-transition-days NONCURRENT_DAYS
--noncurrentversion-transition-storage-class TIERNAME

DAYS - The number of calendar days after which MinIO marks an object as eligible for transition.
```

It turns out that on the NVM storage us need to have the same volume as on the HDD storage in order to avoid overflowing the NVMe storage.

Can you please tell me if we got it right in the documentation?

What happens if the NVMe storage is smaller than the HDD storage and more space is used than is available on NVMe?

**Comment by merwan** [comment]
A quick feedback on our own experience with a two node MinIO setup with 144 TB HDD storage. We do a lot of write operations with small files (less than 500 kb) and we had very poor performance and an over the top IO wait. MinIO struggled to handle a queue with 30000 files during a whole day.
We followed the advice to disable the write cache found here and on https://github.com/minio/minio/issues/17007#issuecomment-1502835913 and the performance changed completely: the queue was handled in less than 5 minutes! We can clearly see the impact on the CPU usage in this graph (configuration changed at 17:55):
![image](https://github.com/user-attachments/assets/75a361eb-9cf7-4751-9e9e-015df7b21bb8)

So thanks for this information which is sort of hidden in the documentation: I only found a reference to disabling the cache here https://min.io/docs/minio/linux/operations/concepts/architecture.html#distributed-minio-deployments
>  MinIO recommends against caching, either at the drive or the controller layer. Either type of caching can cause I/O spikes as the cache fills and clears, resulting in unpredictable performance.

**Reply by harshavardhana** [reply]
Write cache play against you when you have a lot of WRITEs; drive caches cannot handle bursts in writes as they need to dequeue from the cache and they have hardly good logic in there to do so. 

Drive vendors claim features and even enable them by default, which are detrimental in reality.
